guardian/pinboard
# Implement 'mentions' (for individuals)

Co-authored-by: @twrichards 

## What does this change?
This PR adds the functionality to 'mention' individuals in Pinboard:
- load all users from the newly-implemented 'users table' (see #44 and #43) once on page load (this may need to be reconsidered as our number of users grows - when experimenting with thousands of users we observed performance degradation)
- replace the use of the native `textarea` for message input with a React component that allows the user to select a user from the users table to mention, triggered by entering the `@` character. We used [`react-text-area-autocomplete`](https://github.com/webscopeio/react-textarea-autocomplete) for this autocomplete, but it required an update to the type definitions which we did in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51950
- new `mentions` field on Item table (containing email addresses of the people mentioned) and rendering mentions on the item display (red background for mentions of current user, bold for others - for familiarity with Google chat, although will change slightly in future when we implement designs)
- once we had the user lookup on the client, to save storage on the Item table, we replaced the `user` field (a JSON object containing various fields for every Item) with just a `userEmail` field. We then perform lookup on the client to resolve `firstName`, `lastName` and `avatarUrl` for each Item (falling back to displaying just their email if user not present in user table - shouldn't happen)
- [Unfortunately, `AppSync` doesn't support filtering subscriptions on array/list-based fields](https://thespblog.net/blog/appsync-subscriptions/#:~:text=There%20is%20no%20way%20to%20control%20what%20gets%20sent%20through%20subscriptions%20except%20through%20filters%20during%20the%20actual%20connection.), as such we had to create an overall subscription listening for `pinboardId` and `mentions` fields on all `onCreateItem` and then checking on the client whether the mention applies to the current user, and setting the unread flags accordingly
- reinstated the back button for when we have a pinboard pre-selected, but changed the behaviour of the 'select pinboard' view to open the relevant composer article in a new tab with pinboard expanded (to surface mentions on pinboards other than the pre-selected one)
- various small refactors, fixes and tweaks.

## How to test
Easiest to deploy this branch to `CODE`, then: 
- open Grid `TEST`
- in another tab, open a Composer piece which is tracked in Workflow, then open Pinboard and enter the `@` character, and begin typing your name, use the autocomplete to complete your name and hit Enter/'Send'
- observe the unread badge in the Grid tab, click to expand Pinboard, then select the pinboard with the unread badge (should be top of the list), observe your name highlighted with a red background in the message
- still in Grid, open a different pinboard and `@` yourself again
- back in Composer, observe the small unread badge on the 'Back' button, and click it to get back to the list of pinboards, then click on the pinboard with the unread badge, which should open the corresponding Composer piece in a new tab with Pinboard expanded.

## How can we measure success?
Previously, users would have to have the relevant pinboard open, now the pinboard can be brought to their attention when other people need their involvement.

## Images

![mentions](https://user-images.githubusercontent.com/15648334/114019728-aa871e00-9866-11eb-8278-b8c090f01538.gif)
